### PR TITLE
complete all security requirements

### PR DESCRIPTION
**Pull request by esidarapp**

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -79,8 +81,14 @@
 		<dependency>
 			<groupId>org.springdoc</groupId>
 			<artifactId>springdoc-openapi-starter-webmvc-ui</artifactId>
-			<version>2.1.0</version>
+			<version>${org.springdoc.version}</version>
 		</dependency>
+		<dependency>
+			<groupId>org.springframework.boot</groupId>
+			<artifactId>spring-boot-starter-security</artifactId>
+			<version>${org.springframework.boot}</version>
```
Comment (Elena-Bruyako) [inline pom.xml]:
```suggestion
```

File: pom.xml [hunk 2]
```diff
@@ -19,6 +19,8 @@
 		<org.mapstruct.version>1.5.5.Final</org.mapstruct.version>
 		<lombok.mapstruct.binding.version>0.2.0</lombok.mapstruct.binding.version>
 		<mysql.version>8.0.33</mysql.version>
+		<org.springdoc.version>2.1.0</org.springdoc.version>
+		<org.springframework.boot>3.2.4</org.springframework.boot>
```
Comment (Elena-Bruyako) [inline pom.xml]:
```suggestion
```

File: src/main/java/com/example/spring/controller/UserController.java [hunk 3]
```diff
@@ -16,14 +18,20 @@
 @Tag(name = "User Management", description = "Endpoints for managing user authentication")
 @RestController
 @RequiredArgsConstructor
-@RequestMapping("/users")
-public class AuthenticationController {
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/spring/controller/UserController.java]: Revert name please, this controller provides register and login (in next HW) endpoints

File: src/main/java/com/example/spring/controller/UserController.java [hunk 4]
```diff
@@ -16,14 +18,20 @@
...
     @Operation(summary = "Create a new user", description = "Create a new user")
     public UserResponseDto register(
             @RequestBody @Valid UserRegistrationRequestDto request) throws RegisterException {
-        return registerService.register(request);
+        return userService.register(request);
+    }
+
+    @GetMapping
+    @Operation(summary = "Get all users", description = "Get a list of all available users")
+    public List<UserResponseDto> getAllUsers() {
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/spring/controller/UserController.java]: useless endpoint

File: src/main/java/com/example/spring/dto/book/CreateBookRequestDto.java [hunk 5]
```diff
@@ -13,12 +13,14 @@ public class CreateBookRequestDto {
     private String title;
     @NotNull
     private String author;
+    @NotNull
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/spring/dto/book/CreateBookRequestDto.java]: For String use `@NotBlank`

File: src/main/java/com/example/spring/dto/book/UpdateBookRequestDto.java [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+import com.example.spring.validate.annotations.CoverImage;
+import com.example.spring.validate.annotations.Isbn;
+import jakarta.annotation.Nullable;
+import jakarta.validation.constraints.Min;
+import java.math.BigDecimal;
+import lombok.Data;
+
+@Data
+public class UpdateBookRequestDto {
+    @Nullable
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/spring/dto/book/UpdateBookRequestDto.java]:
```suggestion
    @NotBlank
```

check all places with String

File: src/main/java/com/example/spring/model/User.java [hunk 7]
```diff
@@ -28,4 +42,57 @@ public class User {
...
+    @JoinTable(
+            name = "users_roles",
+            joinColumns = @JoinColumn(name = "user_id"),
+            inverseJoinColumns = @JoinColumn(name = "role_id")
+    )
+    private Set<Role> roles;
+
+    @Override
+    public Collection<? extends GrantedAuthority> getAuthorities() {
+        Set<SimpleGrantedAuthority> authorities = new HashSet<>();
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/spring/model/User.java]:
```suggestion
        return roles;
```

File: src/main/java/com/example/spring/model/User.java [hunk 8]
```diff
@@ -28,4 +42,57 @@ public class User {
...
     private String shippingAddress;
+    @Column(nullable = false)
+    private boolean isDeleted;
+    @ManyToMany(fetch = FetchType.EAGER)
+    @JoinTable(
+            name = "users_roles",
+            joinColumns = @JoinColumn(name = "user_id"),
+            inverseJoinColumns = @JoinColumn(name = "role_id")
+    )
+    private Set<Role> roles;
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/spring/model/User.java]:
```suggestion
    private Set<Role> roles = new HashSet<>();

```

File: src/main/java/com/example/spring/security/CustomUserDetailsService.java [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+
+@RequiredArgsConstructor
+@Service
+public class CustomUserDetailsService implements UserDetailsService {
+    private final UserRepository userRepository;
+
+    @Override
+    public UserDetails loadUserByUsername(String email) throws UsernameNotFoundException {
+        return userRepository.findByEmail(email)
+                .orElseThrow(() -> new UsernameNotFoundException(
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/spring/security/CustomUserDetailsService.java]:
```suggestion
                .orElseThrow(() -> new EntityNotFoundException(

```

File: src/main/java/com/example/spring/service/user/UserServiceImpl.java [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+    private final RoleRepository roleRepository;
+
+    @Override
+    public UserResponseDto register(
+            UserRegistrationRequestDto request) throws RegisterException {
+        if (userRepository.existsByEmail(request.getEmail())) {
+            throw new RegisterException("User with email "
+                    + request.getEmail() + " already exists.");
+        }
+        request.setPassword(passwordEncoder.encode(request.getPassword()));
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/spring/service/user/UserServiceImpl.java]:
```suggestion
        User user = userMapper.toModel(request);
        Role userRole = roleRepository.findByName(Role.RoleName.USER)
                .orElseThrow(() -> new EntityNotFoundException("Role USER not found"));
        user.setRoles(Set.of(userRole));
        user.setPassword(passwordEncoder.encode(request.getPassword()));

```

File: src/main/resources/db/changelog/changes/05-add-foreign-keys-to-users-roles-table.yaml [hunk 11]
```diff
@@ -0,0 +1,17 @@
+databaseChangeLog:
+  - changeSet:
+      id: add-foreign-keys-to-users-roles-table
+      author: eugene
+      changes:
+        - addForeignKeyConstraint:
```
Comment (Elena-Bruyako) [inline src/main/resources/db/changelog/changes/05-add-foreign-keys-to-users-roles-table.yaml]: replace FK to create-users-roles-table changelog

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
Good job! Let's improve your solution!

File: src/main/java/com/example/spring/dto/book/UpdateBookRequestDto.java (hunk 6 above)
Comment (esidarapp) [reply src/main/java/com/example/spring/dto/book/UpdateBookRequestDto.java]: I need the string to be possibly null. That`s why i used @Nullable. Probably it is better to remove  this annotion at all.

File: src/main/java/com/example/spring/model/User.java (hunk 7 above)
Comment (esidarapp) [reply src/main/java/com/example/spring/model/User.java]: I don't know where the mistake is, but I tried to follow your suggestions and had only a 403 error.

Comment (Elena-Bruyako) [reply src/main/java/com/example/spring/model/User.java]: need to discuss in chat or QnA. 
Spring Security uses prexif ROLE_, you need to check how roles store in DB, in debugging check endpoint. Maybe Spring add ROLE prefix, but finds in DB only USER without prefix.

File: src/main/java/com/example/spring/controller/UserController.java [hunk 12]
```diff
@@ -16,14 +16,14 @@
 @Tag(name = "User Management", description = "Endpoints for managing user authentication")
 @RestController
 @RequiredArgsConstructor
-@RequestMapping("/users")
-public class AuthenticationController {
-    private final RegisterService registerService;
+@RequestMapping("/auth")
+public class UserController {
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/spring/controller/UserController.java]:
```suggestion
public class AuthenticationController {
```

File: src/main/java/com/example/spring/dto/book/UpdateBookRequestDto.java [hunk 13]
```diff
@@ -0,0 +1,20 @@
+package com.example.spring.dto.book;
+
+import com.example.spring.validate.annotations.CoverImage;
+import com.example.spring.validate.annotations.Isbn;
+import jakarta.validation.constraints.Min;
+import java.math.BigDecimal;
+import lombok.Data;
+
+@Data
+public class UpdateBookRequestDto {
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/spring/dto/book/UpdateBookRequestDto.java:10]: You can use CreateBookRequestDto to update book, the fields are the same

File: src/main/java/com/example/spring/controller/BookController.java [hunk 14]
```diff
@@ -30,38 +32,44 @@
 public class BookController {
     private final BookService bookService;
 
+    @PreAuthorize("hasRole('ROLE_USER')")
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/spring/controller/BookController.java:35]:
```suggestion
    @PreAuthorize("hasRole('USER')")
```
check other places

File: src/main/java/com/example/spring/model/Role.java [hunk 15]
```diff
@@ -0,0 +1,39 @@
...
+    @Enumerated(EnumType.STRING)
+    private RoleName name;
+
+    @Override
+    public String getAuthority() {
+        return name.name();
+    }
+
+    public enum RoleName {
+        USER,
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/spring/model/Role.java]:
```suggestion
        ROLE_USER,
        ROLE_ADMIN
```

File: src/main/java/com/example/spring/model/User.java [hunk 16]
```diff
@@ -28,4 +42,56 @@ public class User {
...
+    @JoinTable(
+            name = "users_roles",
+            joinColumns = @JoinColumn(name = "user_id"),
+            inverseJoinColumns = @JoinColumn(name = "role_id")
+    )
+    private Set<Role> roles = new HashSet<>();
+
+    @Override
+    public Collection<? extends GrantedAuthority> getAuthorities() {
+        Set<SimpleGrantedAuthority> authorities = new HashSet<>();
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/spring/model/User.java]:
```suggestion
        return roles;
```

File: src/main/java/com/example/spring/service/user/UserServiceImpl.java [hunk 17]
```diff
@@ -0,0 +1,47 @@
...
+        User user = userMapper.toModel(request);
+        Role userRole = roleRepository.findByName(Role.RoleName.USER)
+                .orElseThrow(() -> new EntityNotFoundException("Role USER not found"));
+        user.setRoles(Set.of(userRole));
+        user.setPassword(passwordEncoder.encode(request.getPassword()));
+        return userMapper.toDto(userRepository.save(user));
+    }
+
+    @Override
+    public List<UserResponseDto> findAllUsers() {
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/spring/service/user/UserServiceImpl.java]: remove

File: src/main/java/com/example/spring/service/user/UserService.java [hunk 18]
```diff
@@ -3,7 +3,10 @@
 import com.example.spring.dto.user.UserRegistrationRequestDto;
 import com.example.spring.dto.user.UserResponseDto;
 import com.example.spring.exception.RegisterException;
+import java.util.List;
 
-public interface RegisterService {
+public interface UserService {
     UserResponseDto register(UserRegistrationRequestDto request) throws RegisterException;
+
+    List<UserResponseDto> findAllUsers();
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/spring/service/user/UserService.java]: remove

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see comments

File: src/main/java/com/example/spring/dto/book/UpdateBookRequestDto.java (hunk 13 above)
Comment (esidarapp) [reply src/main/java/com/example/spring/dto/book/UpdateBookRequestDto.java:10]: Yeah, but i need all fields not null to create a book, but for update its not optimal to fill all fields in json. That's why i decided to create another DTO for updating book.

**Review by Elena-Bruyako — APPROVED** [review]
Well done

**Pull request merged**